pkg/boot/util/transport.go

```diff
@@ -0,0 +1,53 @@
+package util
+
+import "net"
+
+type DialFunc func(net.Addr) (net.PacketConn, error)
+
+type ListenFunc func(net.Addr) (net.PacketConn, error)
+
+type Transport struct {
+	Dial   DialFunc
+	Listen ListenFunc
+}
+
+var (
+	MulticastTransport = Transport{Dial: UdpDial, Listen: MulticastListen}
+	UdpTransport       = Transport{Dial: UdpDial, Listen: UdpListen}
+)
+
+func UdpDial(addr net.Addr) (net.PacketConn, error) {
```

Comment:
**Minor point:**  the Go naming convention would be `DialUDP`, `ListenUDP`, etc.

**Major point (API design):**  I'm skeptical about providing the same interface for survey and crawl, because their underlying semantics are quite different:

- Survey is multicast, so it requires a separate socket for read/write.  This is why it makes sense to pass in a pair of constructors via `Transport`.
- Crawl is unicast and only requires a single socket for read/write.  As such, it makes more sense to pass in a fully-constructed `net.PacketConn`.  I'm not sure what the extra level of indirection provides.

My concern is that trying to unite these under a common `Transport` type will produce a leaky abstraction.  Users will need to know low-level details about how crawl vs survey use the transport type internally.  For instance:  they will need to understand why passing `MulticastTransport` to `crawl.New` is absurd.  Better to guard against this kind of semantic confusion via type-checking.

My suggestions is:
- Accept `net.PacketConn` in `crawl.New`.
- Accept `survey.Transport` in `survey.New`.
- Remove `util` subpackage

Reply:
The thing is that `Crawl` uses a different connection for receiving requests and sending them. This helps to simplify the protocol since we avoid having to filter and differentiate request and response messages.

Moreover, do not forget that since we created the `Transport` constructor, the user is not forced to use `Multicasting` for the `Surveyor`. Therefore, in the end, the logic of having `Dial` and `Listen` functions makes sense for both cases similarly. Since, the user still needs to understand why `Survey` should only receive a `MulticastTransport`.

I think that if we don't want to unify both abstractions, then, we should redesign the transport abstraction and force user to a multicast / point-to-point communication. This should be done by not allowing to pass a connection, but limiting the user to specifying the connection type as `udp4` / `udp6`.